tools/linter/adapters/stable_shim_usage_linter.py

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python3
+"""
+STABLE_SHIM_USAGE: Ensures that calls to versioned shim functions from
+torch/csrc/stable/c/shim.h in torch/csrc/stable are properly wrapped in
+TORCH_FEATURE_VERSION macros corresponding to the version when those
+functions were introduced.
+"""
+
+from __future__ import annotations
+
+import argparse
+import json
+import logging
+import re
+import sys
+from enum import Enum
+from pathlib import Path
+from typing import NamedTuple
+
+
+LINTER_CODE = "STABLE_SHIM_USAGE"
+
+
+class PreprocessorTracker:
+    """
+    Helper class to track preprocessor directives and version blocks.
+
+    This class maintains state as it processes C/C++ preprocessor directives
+    (#if, #elif, #else, #endif) and tracks which code is inside version blocks.
+    """
+
+    def __init__(self):
+        """Initialize the preprocessor tracker."""
+        # Stack of (is_version_block, version_tuple) tuples
+        # is_version_block: True if this is a TORCH_FEATURE_VERSION >= TORCH_VERSION_X_Y_0 block
+        # version_tuple: (major, minor) if is_version_block is True, else None
+        self.preprocessor_stack: list[tuple[bool, tuple[int, int] | None]] = []
+
+        # Current version requirement (if inside a version block)
+        self.current_version: tuple[int, int] | None = None
```

Comment:
nit: maybe call this version_of_block cuz current_version got conflated for me with the current torch version (like 2.10)
```suggestion
        self.version_of_block: tuple[int, int] | None = None
```